File: desktop/src/services/processMemory.ts
import { invoke } from '@tauri-apps/api/core'

/**
 * Process memory snapshot returned by the `browser_process_memory` command.
 * Mirrors the Rust struct in `src-tauri/src/process_memory.rs`.
 */
export interface ProcessMemorySnapshot {
  /** Working-set (resident) size in bytes. */
  workingSetBytes: number
  /** Peak working-set size in bytes since process start. */
  peakWorkingSetBytes: number
  /** Committed / pagefile usage in bytes. */
  commitBytes: number
  /** Cumulative page-fault count since process start. */
  pageFaultCount: number
}

const isTauri = () => '__TAURI_INTERNALS__' in window

/**
 * Snapshot the current process memory accounting from the Rust side.
 *
 * Returns `null` outside Tauri or when the host platform does not yet
 * implement the query (the frontend renders an em-dash in that case).
 */
export async function readProcessMemory(): Promise<ProcessMemorySnapshot | null> {
  if (!isTauri()) return null
  try {
    return await invoke<ProcessMemorySnapshot | null>('browser_process_memory')
  } catch {
    return null
  }
}

/** Human-friendly byte size with adaptive units. */
export function formatBytes(bytes: number | null | undefined): string {
  if (bytes == null) return '—'
  if (bytes < 1024) return `${bytes} B`
  const units = ['KB', 'MB', 'GB', 'TB']
  let value = bytes / 1024
  let unitIndex = 0
  while (value >= 1024 && unitIndex < units.length - 1) {
    value /= 1024
    unitIndex += 1
  }
  const fixed = value >= 100 ? value.toFixed(0) : value >= 10 ? value.toFixed(1) : value.toFixed(2)
  return `${fixed} ${units[unitIndex]}`
}

File: desktop/src-tauri/src/process_memory.rs
//! Process memory snapshot for the resource panel.
//!
//! On Windows uses `K32GetProcessMemoryInfo` from `psapi.dll` against the
//! current process. Other targets return `None` so the frontend degrades to
//! "—" without breaking.

#[cfg(target_os = "windows")]
mod imp {
    use serde::Serialize;
    use windows_sys::Win32::System::ProcessStatus::{
        K32GetProcessMemoryInfo, PROCESS_MEMORY_COUNTERS,
    };
    use windows_sys::Win32::System::Threading::GetCurrentProcess;

    #[derive(Debug, Serialize)]
    pub struct ProcessMemory {
        /// Resident / working-set size in bytes (current physical memory in use).
        pub working_set_bytes: u64,
        /// Peak working-set size in bytes.
        pub peak_working_set_bytes: u64,
        /// Virtual size committed in bytes.
        pub commit_bytes: u64,
        /// Page-fault count (informational).
        pub page_fault_count: u64,
    }

    pub fn sample() -> Option<ProcessMemory> {
        let mut counters: PROCESS_MEMORY_COUNTERS = unsafe { std::mem::zeroed() };
        let mut out_size = std::mem::size_of::<PROCESS_MEMORY_COUNTERS>() as u32;
        let ok = unsafe {
            K32GetProcessMemoryInfo(
                GetCurrentProcess(),
                &mut counters,
                out_size,
            )
        };
        if ok == 0 {
            return None;
        }
        Some(ProcessMemory {
            working_set_bytes: counters.WorkingSetSize as u64,
            peak_working_set_bytes: counters.PeakWorkingSetSize as u64,
            commit_bytes: counters.PagefileUsage as u64,
            page_fault_count: counters.PageFaultCount as u64,
        })
    }
}

#[cfg(not(target_os = "windows"))]
mod imp {
    use serde::Serialize;

    #[derive(Debug, Serialize)]
    pub struct ProcessMemory {
        pub working_set_bytes: u64,
        pub peak_working_set_bytes: u64,
        pub commit_bytes: u64,
        pub page_fault_count: u64,
    }

    pub fn sample() -> Option<ProcessMemory> {
        None
    }
}

pub use imp::{sample, ProcessMemory};

/// Tauri command: snapshot of the current process memory accounting.
///
/// Returns `None` on platforms where the OS query is not yet implemented
/// (frontend should render an em-dash instead of zero).
#[tauri::command]
pub fn browser_process_memory() -> Option<ProcessMemory> {
    sample()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn returns_shape_when_supported() {
        let result = sample();
        if let Some(snapshot) = result {
            assert!(snapshot.peak_working_set_bytes >= snapshot.working_set_bytes.min(u64::MAX));
        }
    }
}